cpp/src/utilities/type_dispatcher.hpp

```diff
@@ -16,20 +16,216 @@
+    using type = no_invocation_operator_return_type;
+};
+
+template <class F, typename T, typename... Us>
+struct invocation_operator_return_type_inner<F, T, true, Us...> {
+    using type = decltype(std::declval<F>().template operator()<T>(std::declval<Us>()... ));
+};
+
+/**
+ * The return type of a functor's first (templated) `operator()` which is actually defined
+ * (or a dummy type otherwise)
+ */
+template <class F, typename T, typename... Us>
+using invocation_operator_return_type =
+    typename invocation_operator_return_type_inner<F, T, has_templated_invoke_operator<F, T, Us...>::value, Us...>::type;
+
+template <class F, typename... Us>
+struct fnm_helper {
+
+    template <typename T>
+    using iort = invocation_operator_return_type<F, T, Us...>;
+
+    template <typename... List>
+    using transformed = typename first_non_matching<
+        no_invocation_operator_return_type,
+        iort<List>...>::type;
+};
+
+} // namespace detail
+
+// As you read the above code (after the trait), you might be wondering:
+//
+// 1. Why is invoke_or_fail a struct, rather than a function? After all, the only thing we do
+//    with it is invoke it (with operator())?
+// 2. Why have two structs, an internal one (invoke_or_fail_inner) and an external one (invoke_or_fail)?
+//    Isn't the internal one enough?
+//
+// Answers:
+//
+// 1. A C++ limitation. We could not have performed the tag dispatching with just functions, as
+//    C++14 does not allow this kind of partial specialization for freestanding functions.
+// 2. Note the template parameter on the outer struct. It allows client code to specify explicitly
+//    which type it's interested in for instantiation of F's templated operator(), and
+//    specify all other template parameters explicitly via the call to the outer struct's operator().
+//    See the (only?) example of use in the type_dispatcher mechanism, below.
+//
+// Finally - Note that some or most of this ugly TMP can easily go away if C++17 or C++20 is available to
+// us - using if constexpr to avoid SFINAE, and eventually using variadic lambdas which may eliminate
+// the type_dispatcher altogether. Still, even with C++14 it's possible the
+
+/**---------------------------------------------------------------------------*
+ * @brief Maps a C++ type to it's corresponding gdf_dtype.
+ *
+ * When explicitly passed a template argument of a given type, returns the
+ * appropriate `gdf_dtype` for the specified C++ type.
+ *
+ * For example:
+ * ```
+ * return gdf_dtype_of<int32_t>();        // Returns GDF_INT32
+ * return gdf_dtype_of<cudf::category>(); // Returns GDF_CATEGORY
+ * ```
+ *
+ * @tparam T The type to map to a `gdf_dtype`
+ *---------------------------------------------------------------------------**/
+template <typename T>
+inline constexpr gdf_dtype gdf_dtype_of() {  return GDF_invalid; };
+
+template <> inline constexpr gdf_dtype gdf_dtype_of< int8_t            >() { return GDF_INT8;            };
```

Comment:
I like this simplification / compaction of `gdf_dtype_of`

Reply:
I pushed the missing pragma. About your general point: 

* With your example, you're forcing API function implementers to use some TMP themselves. With my PR, they don't have to - they can implement for the types they want. It's true that the TMP I use is more complicated than the enable_if and not enable_if in your example, but - it's only written once, centrally. Also, It does not require maintenance, per se. That is, it doesn't need to change to the addition or removal of gdf types. Only if we decide to change how the type dispatcher works will this need to change.
* It's true that with the example you gave, the error message is more specific, but a message about an unsupported dtype is kind of ok as an error message. The type name will be part of the stack, so when debugging this will be visible to the developer. However, if you like, I could also introduce a mechanism for the developer to write a custom error message explaining why certain types are supported and others aren't.

This PR is not ultra-significant. I just needed/wanted something like; and Jake said he would like to see it too. I still think it's worth it, but if you decide otherwise I'll drop this PR and create a separate one for the "compaction" of `gdf_dtype_of`.

Reply:
I agree with Mark on this one.

The idea of what you proposed initially appealed to me. However, forcing developers to explicitly specialize with `enable_if` may be extra work for the developer, but I believe that will lead to more useful error messages for the end users---a majority of whom are Python users that can't/won't dig into the C++ source to find out the cause of the error.

Reply:
Ok. so - closing this PR for now.